lib/node_modules/@stdlib/string/next-grapheme-cluster-break/lib/main.js

```diff
@@ -65,13 +69,20 @@ function nextBreak( str, idx ) {
 	var mid;
 	var i;
 
+	if ( !isString( str ) ) {
+		throw new TypeError( 'invalid argument. Must provide a string. Value: `' + str + '`.' );
+	}
 	len = str.length;
-	if ( idx < 0 ) {
+	if ( len === 0 ) {
 		return 0;
 	}
-	if ( idx >= len - 1 ) {
-		return len;
+	if ( !isNonNegativeInteger( idx ) ) {
+		throw new TypeError( 'invalid argument. Must provide a valid position (nonnegative integer). Value: `' + idx + '`.' );
+	}
+	if ( idx >= len ) {
```

Comment:
Would it make sense to simply return `-1` rather than throw an error? Otherwise, as a user, I need to perform a sanity check before providing a position argument. What is the behavior for, e.g., `String#indexOf`?

```
In [61]: 'foo'.indexOf('o')
Out[61]: 1

In [62]: 'foo'.indexOf('o',1)
Out[62]: 1

In [63]: 'foo'.indexOf('o',2)
Out[63]: 2

In [64]: 'foo'.indexOf('o',3)
Out[64]: -1
```

I think we should follow this precedent where we don't throw if provided an index which exceeds the last string index.

Reply:
I think `indexOf` and `nextGraphemeClusterBreak` are different in the sense that their range of output varies: `indexOf` will always return on the set `[0, str.length - 1]` (apart from "error" values in which case `-1` is used) whereas `nextGraphemeClusterBreak` returns `str.length` in case the `fromIndex` is greater than or equal to the last break within bounds but strictly less than `str.length` (a case which may happen frequently especially if the last grapheme cluster consists of multiple code points).
If we return `-1` here instead of throwing, what do we in case of `prevGraphemeClusterBreak`? There `-1` has a special meaning: in case the `beforeIndex` is less than or equal to the last (viewing backwards) break within bounds but greater than or equal to `0`, `-1` is returned (again, this case might happen quite frequently).

Reply:
As an alternative to `-1`, we could return `null`. This would satisfy both `next*` and `prev*` constraints. This has the problem of multiple possible return types, which introduces a potential perf cliff and requires users to account for multiple different return values. We could also return `NaN` or some other sentinel value.

I suppose my main hangup here is whether we should be unforgiving if a user provides a `fromIndex` which is out-of-bounds. Prior precedent suggests that we should be forgiving, and not throw, but we're confronted here with possible inconsistencies between `next*` and `prev*` if we follow precedent.

Just so I understand things correctly, the return value equals

-   `str.length`: if unable to find a cluster break when searching start-to-end via `next*`.
-   `-1`: if unable to find a cluster break when searching end-to-start via `prev*`.
-   `Error`: if provided an out-of-bounds index.
-   `[0, str.length-1]`: if able to find a cluster break when searching either start-to-end or end-to-start.

From the above, not clear to me why necessary to return `str.length` if unable to find a cluster break when searching start-to-end via `next*`. We'd be returning an out-of-bounds string index. Why return `str.length` and not `-1`, when `-1` is the conventional value for indicating that something is not found? I don't think I am understanding the advantage of `str.length` in this scenario.

If we chose to return `-1` if unable to find, then we shouldn't be concerned about `-1` holding a special meaning for `prev*`. Similarly, we'd just return `-1` if unable to find. That we had to iterate from end-to-start is not material.

Maybe you can clarify the design decision here. :)

Reply:
Just to be clear, the commit [238f728](https://github.com/stdlib-js/stdlib/pull/361/commits/238f72899b98a052d911034425c4039b561d73a4) extends support for negative arguments now. We throw `Error` only in case `fromIndex >= str.length`. But otherwise, your understanding is correct.

Re: why necessary to return `str.length` in case of `next*`
This is because there are at least 2 distinct scenarios in which we wouldn't find a cluster break:

1. If we provide `fromIndex >= str.length`, in which case we are throwing `Error`.
2. Imagine that the last grapheme cluster has many codepoints. Something like this where "|" means a "conventional" break (sorry for the crude diagram) - `a | bc | defgh`. Here if we give the `fromIndex` anything between `[3, 7]`, we wouldn't know that this is the last grapheme cluster if we'd be returning `-1` for this and the first case above. And hence currently `str.length` is returned.

So `str.length` means that we have an extra break beyond the "conventional" breaks at the end of the string for `next*`, and by extension (the first grapheme cluster can also have many codepoints), we have an extra break at the start of the string for `prev*` where we return `-1`.

In the end it comes down to distinguishing between these 2 cases for both `next*` and `prev*`.

Reply:
If we give a `fromIndex` on the closed interval `[3,7]`, wouldn't the last grapheme cluster be implied if we returned `-1`?

Okay...I think I see your point. From the POV of a user, if I receive a return value of `-1`, I won't be able to distinguish between either having provided a `fromIndex` exceeding the string length or having been unable to locate a grapheme cluster break. But why is this important from the POV of a user?

If I provide a `fromIndex` exceeding the string length and get back a `-1`, I know that a cluster break does not occur after the provided index. Similarly if `fromIndex` is within bounds and we're unable to find another break, we can conclude that a break does not occur after the provided index.

In what scenario do I care that I'm unable to distinguish whether I've provided a `fromIndex` exceeding the string length and another in which we're unable to find a break?

Reply:
It seems odd to me that we'd return a positive index value which exceeds a string's index bounds (either `str.length` or `-1`) to indicate that a break occurs somewhere beyond string bounds. But quite possible that I am not understanding the use cases sufficiently.

Reply:
I suppose the more general point I am getting at is: why do I, as a user, care whether I am able to distinguish between providing an out-of-bounds index vs being unable to find a cluster break **within** the string? What would I do with that extra information?

Reply:
The implementation for `num-grapheme-clusters` [here](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/string/num-grapheme-clusters/lib/main.js#L64) is dependent on this assumption that the next break after any `fromIndex` on the interval `[3, 7]` would be `str.length`. That implementation can obviously be tweaked to satisfy the new constraints, you are certainly right that we can infer that we have reached the last cluster if we combine the 2 facts that the return value is `-1` and `fromIndex < str.length`. 

From an user's perspective as well, the only thing that the current implementation does is save her from having to make this inference on their end.

Reply:
I think my main worry with throwing is that I, as a user, need to do

```javascript
function getBreak( str, fromIndex ) {
    try {
        return nextClusterBreak( str, fromIndex );
    } catch ( err ) {
        return null; // or some other value
    }
}
```

or

```javascript
function getBreak( str, fromIndex ) {
    if ( fromIndex >= str.length ) {
        return null; // or some other value
    }
    return nextClusterBreak( str, fromIndex );
}
```

Meaning, if I develop a library using this package, I need to engage in defensive behaviors to prevent an exception being thrown when provided, e.g., a string provided to me, not necessarily one that I have supplied (e.g., an input textbox).

I suppose my mental model is strongly anchored in the `indexOf` way of things, and my concern is that users will be, as well.

For the code you refer to above, seems like this to be written

```javascript
brk = nextBreak( str, idx );
while ( brk !== -1 ) {
	count += 1;
	idx = brk;
	brk = nextBreak( str, idx );
}
``` 

which would be a relatively minor change, it would seem.

Reply:
Regarding distinguishing the two cases, that also applies to `String#indexOf`. I'm unable to distinguish whether I've provided an out-of-bounds index or I was unable to find my search value. I have to do the check you describe above. Namely, `idx === -1 && fromIndex < str.length`.

Reply:
Oh, then I guess this fits perfectly in that precedent. And you are right, that code change for `num-grapheme-clusters` is straightforward.
I'll admit that I never gave it a thought from a third party input perspective which may not be in our control.

I'll do these changes along with making the `fromIndex` argument optional later today.

Reply:
Awesome! Thank you for taking the time to explain your thinking! Certainly helped me think through things. Really appreciate it. :)